File: projects/dejavu.ts/dejavu-language/src/parser.ts
import {
    DEFAULT_LANGUAGE,
    type IrDocument,
    type IrNode,
    type Language,
    type ParseOptions,
    type Trim,
} from "@dejavu/types";
import { ParseError } from "./error";
import { parseExpr } from "./expr";
import { lexCode } from "./lexer";
import type { CodeToken } from "./token";

type TrimMode = Trim;

/** Classify a code block from its token stream (no string-prefix hacks). */
type CodeHead =
    | { kind: "if"; testSlice: string; testAbs: number }
    | { kind: "loop"; item: string; iterSlice: string; iterAbs: number }
    | { kind: "else_if"; testSlice: string; testAbs: number }
    | { kind: "else" }
    | { kind: "end_if" }
    | { kind: "end_loop" }
    | { kind: "extends"; parentSlice: string; parentAbs: number }
    | { kind: "include"; pathSlice: string; pathAbs: number }
    | { kind: "block"; name: string }
    | { kind: "end_block" }
    | { kind: "super" }
    | { kind: "expr"; exprSlice: string; exprAbs: number };

function sliceFromTokens(
    content: string,
    contentBase: number,
    tokens: CodeToken[],
): { slice: string; abs: number } {
    if (tokens.length === 0) return { slice: "", abs: contentBase };
    const start = tokens[0]!.start;
    const end = tokens.at(-1)!.end;
    return { slice: content.slice(start, end), abs: contentBase + start };
}

function classifyCode(content: string, contentBase: number, tokens: CodeToken[]): CodeHead {
    if (tokens.length === 0) {
        return { kind: "expr", exprSlice: "", exprAbs: contentBase };
    }
    const t0 = tokens[0]!;
    if (t0.kind === "Ident" && t0.text === "if") {
        const { slice, abs } = sliceFromTokens(content, contentBase, tokens.slice(1));
        return { kind: "if", testSlice: slice, testAbs: abs };
    }
    if (t0.kind === "Ident" && t0.text === "loop") {
        if (tokens[1]?.kind !== "Ident") {
            throw new ParseError("loop requires item identifier", {
                start: contentBase + t0.start,
                length: Math.max(1, (tokens.at(-1)?.end ?? t0.end) - t0.start),
            });
        }
        if (tokens[2]?.kind !== "In") {
            throw new ParseError("loop requires `in`", {
                start: contentBase + t0.start,
                length: Math.max(1, (tokens.at(-1)?.end ?? t0.end) - t0.start),
                label: "expected `item in iterable`",
            });
        }
        const item = tokens[1]!.text;
        const { slice, abs } = sliceFromTokens(content, contentBase, tokens.slice(3));
        return { kind: "loop", item, iterSlice: slice, iterAbs: abs };
    }
    if (t0.kind === "Ident" && t0.text === "extends") {
        const { slice, abs } = sliceFromTokens(content, contentBase, tokens.slice(1));
        if (!slice) {
            throw new ParseError("extends requires a parent template expression", {
                start: contentBase + t0.start,
                length: Math.max(1, t0.end - t0.start),
            });
        }
        return { kind: "extends", parentSlice: slice, parentAbs: abs };
    }
    if (t0.kind === "Ident" && t0.text === "include") {
        const { slice, abs } = sliceFromTokens(content, contentBase, tokens.slice(1));
        if (!slice) {
            throw new ParseError("include requires a path expression", {
                start: contentBase + t0.start,
                length: Math.max(1, t0.end - t0.start),
            });
        }
        return { kind: "include", pathSlice: slice, pathAbs: abs };
    }
    if (t0.kind === "Ident" && t0.text === "block") {
        if (tokens[1]?.kind !== "Ident") {
            throw new ParseError("block requires a name", {
                start: contentBase + t0.start,
                length: Math.max(1, (tokens.at(-1)?.end ?? t0.end) - t0.start),
            });
        }
        if (tokens.length !== 2) {
            throw new ParseError("unexpected tokens after block name", {
                start: contentBase + tokens[2]!.start,
                length: 1,
            });
        }
        return { kind: "block", name: tokens[1]!.text };
    }
    if (t0.kind === "Ident" && t0.text === "super" && tokens.length === 1) {
        return { kind: "super" };
    }
    if (t0.kind === "Ident" && t0.text === "else") {
        if (tokens[1]?.kind === "Ident" && tokens[1].text === "if") {
            const { slice, abs } = sliceFromTokens(content, contentBase, tokens.slice(2));
            return { kind: "else_if", testSlice: slice, testAbs: abs };
        }
        if (tokens.length === 1) return { kind: "else" };
        throw new ParseError("unexpected tokens after `else`", {
            start: contentBase + tokens[1]!.start,
            length: 1,
        });
    }
    if (t0.kind === "Ident" && t0.text === "end") {
        if (tokens[1]?.kind === "Ident" && tokens[1].text === "if" && tokens.length === 2) {
            return { kind: "end_if" };
        }
        if (tokens[1]?.kind === "Ident" && tokens[1].text === "loop" && tokens.length === 2) {
            return { kind: "end_loop" };
        }
        if (tokens[1]?.kind === "Ident" && tokens[1].text === "block" && tokens.length === 2) {
            return { kind: "end_block" };
        }
        throw new ParseError("expected `end if`, `end loop`, or `end block`", {
            start: contentBase + t0.start,
            length: Math.max(1, (tokens.at(-1)?.end ?? t0.end) - t0.start),
        });
    }
    const { slice, abs } = sliceFromTokens(content, contentBase, tokens);
    return { kind: "expr", exprSlice: slice, exprAbs: abs };
}

/** Parse template source into Dejavu IR (T1 + inheritance surface). */
export function parseToIr(source: string, options: ParseOptions | string = {}): IrDocument {
    const opts: ParseOptions = typeof options === "string" ? { file: options } : (options ?? {});
    const file = opts.file ?? "template.dejavu";
    const language = structuredClone(opts.language ?? DEFAULT_LANGUAGE);
    return {
        irVersion: "1.0",
        language,
        body: {
            type: "Template",
            children: new TemplateParser(source, file, language).parse(),
        },
    };
}

class TemplateParser {
    private readonly codeStart: string;
    private readonly codeEnd: string;
    private readonly commentStart: string;
    private readonly commentEnd: string;

    constructor(
        private readonly source: string,
        private readonly file: string,
        language: Language,
    ) {
        this.codeStart = language.template.codeStart;
        this.codeEnd = language.template.codeEnd;
        this.commentStart = language.template.commentStart;
        this.commentEnd = language.template.commentEnd;
    }

    parse(): IrNode[] {
        const [children] = this.parseBody(0, []);
        return children;
    }

    private parseBody(i: number, stop: CodeHead["kind"][]): [IrNode[], number] {
        const children: IrNode[] = [];
        const source = this.source;
        const cs = this.codeStart;

        while (i < source.length) {
            if (source.startsWith(cs, i) && stop.length > 0) {
                const block = this.readCodeBlock(i);
                const head = classifyCode(block.content, block.contentBase, block.tokens);
                if (stop.includes(head.kind)) return [children, i];
            }

            if (source.startsWith(this.commentStart, i)) {
                const end = this.findDelimiter(i + this.commentStart.length, this.commentEnd);
                if (end < 0) {
                    throw new ParseError("unclosed comment", {
                        file: this.file,
                        start: i,
                        length: this.commentStart.length,
                        label: "comment starts here",
                    });
                }
                children.push({
                    type: "Comment",
                    value: source.slice(i + this.commentStart.length, end),
                });
                i = end + this.commentEnd.length;
                continue;
            }

            // Escape: codeStart + "!" → literal codeStart (e.g. `<%!` → `<%`)
            if (source.startsWith(cs + "!", i)) {
                children.push({ type: "Text", value: cs });
                i += cs.length + 1;
                continue;
            }

            if (source.startsWith(cs, i)) {
                const open = i;
                const block = this.readCodeBlock(i);
                i = block.next;
                const head = classifyCode(block.content, block.contentBase, block.tokens);

                if (stop.length > 0 && stop.includes(head.kind)) {
                    return [children, open];
                }

                if (head.kind === "if") {
                    const [node, ni] = this.parseIf(i, head, block.trim);
                    children.push(node);
                    i = ni;
                } else if (head.kind === "loop") {
                    const [node, ni] = this.parseLoop(i, head, block.trim);
                    children.push(node);
                    i = ni;
                } else if (head.kind === "block") {
                    const [node, ni] = this.parseBlock(i, head, block.trim);
                    children.push(node);
                    i = ni;
                } else if (head.kind === "extends") {
                    children.push({
                        type: "Stmt.Extends",
                        parent: parseExpr(head.parentSlice, {
                            source,
                            file: this.file,
                            base: head.parentAbs,
                        }),
                        trim: block.trim,
                    });
                } else if (head.kind === "include") {
                    children.push({
                        type: "Stmt.Include",
                        path: parseExpr(head.pathSlice, {
                            source,
                            file: this.file,
                            base: head.pathAbs,
                        }),
                        trim: block.trim,
                    });
                } else if (head.kind === "super") {
                    children.push({ type: "Stmt.Super", trim: block.trim });
                } else if (
                    head.kind === "end_if" ||
                    head.kind === "end_loop" ||
                    head.kind === "end_block" ||
                    head.kind === "else" ||
                    head.kind === "else_if"
                ) {
                    if (stop.length === 0) {
                        throw new ParseError(`unexpected control \`${head.kind}\``, {
                            file: this.file,
                            start: block.contentBase,
                            length: 1,
                        });
                    }
                    return [children, open];
                } else {
                    children.push({
                        type: "Interpolation",
                        expression: parseExpr(head.exprSlice, {
                            source,
                            file: this.file,
                            base: head.exprAbs,
                        }),
                        trim: block.trim,
                    });
                }
                continue;
            }

            const next = this.nextMarkup(i);
            if (next === null) {
                children.push({ type: "Text", value: source.slice(i) });
                break;
            }
            if (next > i) {
                children.push({ type: "Text", value: source.slice(i, next) });
                i = next;
            } else {
                i += 1;
            }
        }

        return [children, i];
    }

    private parseIf(
        i: number,
        head: Extract<CodeHead, { kind: "if" }>,
        trim: TrimMode,
    ): [IrNode, number] {
        const test = parseExpr(head.testSlice, {
            source: this.source,
            file: this.file,
            base: head.testAbs,
        });
        let [consequent, pos] = this.parseBody(i, ["else_if", "else", "end_if"]);
        i = pos;
        const elseIfs: IrNode[] = [];
        let alternate: IrNode[] | undefined;

        for (;;) {
            const block = this.readCodeBlock(i);
            const h = classifyCode(block.content, block.contentBase, block.tokens);
            if (h.kind === "else_if") {
                i = block.next;
                const t = parseExpr(h.testSlice, {
                    source: this.source,
                    file: this.file,
                    base: h.testAbs,
                });
                const [body, n] = this.parseBody(i, ["else_if", "else", "end_if"]);
                elseIfs.push({ type: "Stmt.ElseIf", test: t, consequent: body, trim: "none" });
                i = n;
            } else if (h.kind === "else") {
                i = block.next;
                const [body, n] = this.parseBody(i, ["end_if"]);
                alternate = body;
                i = n;
                const end = this.readCodeBlock(i);
                if (classifyCode(end.content, end.contentBase, end.tokens).kind !== "end_if") {
                    throw new ParseError("expected `end if`", {
                        file: this.file,
                        start: end.contentBase,
                        length: 1,
                    });
                }
                i = end.next;
                break;
            } else if (h.kind === "end_if") {
                i = block.next;
                break;
            } else {
                throw new ParseError(`expected if closer, got \`${h.kind}\``, {
                    file: this.file,
                    start: block.contentBase,
                    length: 1,
                });
            }
        }

        return [
            {
                type: "Stmt.If",
                test,
                consequent,
                elseIfs,
                ...(alternate ? { alternate } : {}),
                trim,
            },
            i,
        ];
    }

    private parseLoop(
        i: number,
        head: Extract<CodeHead, { kind: "loop" }>,
        trim: TrimMode,
    ): [IrNode, number] {
        const iterable = parseExpr(head.iterSlice, {
            source: this.source,
            file: this.file,
            base: head.iterAbs,
        });
        const [body, pos] = this.parseBody(i, ["end_loop"]);
        i = pos;
        const end = this.readCodeBlock(i);
        if (classifyCode(end.content, end.contentBase, end.tokens).kind !== "end_loop") {
            throw new ParseError("expected `end loop`", {
                file: this.file,
                start: end.contentBase,
                length: 1,
            });
        }
        return [{ type: "Stmt.For", item: head.item, iterable, body, trim }, end.next];
    }

    private parseBlock(
        i: number,
        head: Extract<CodeHead, { kind: "block" }>,
        trim: TrimMode,
    ): [IrNode, number] {
        const [body, pos] = this.parseBody(i, ["end_block"]);
        i = pos;
        const end = this.readCodeBlock(i);
        if (classifyCode(end.content, end.contentBase, end.tokens).kind !== "end_block") {
            throw new ParseError("expected `end block`", {
                file: this.file,
                start: end.contentBase,
                length: 1,
            });
        }
        return [{ type: "Stmt.Block", name: head.name, body, trim }, end.next];
    }

    private readCodeBlock(i: number): {
        tokens: CodeToken[];
        content: string;
        contentBase: number;
        trim: TrimMode;
        next: number;
    } {
        const source = this.source;
        const cs = this.codeStart;
        const ce = this.codeEnd;
        if (!source.startsWith(cs, i)) {
            throw new ParseError(`expected code open \`${cs}\``, {
                file: this.file,
                start: i,
                length: 1,
            });
        }
        let j = i + cs.length;
        let trim: TrimMode = "none";
        const mod = source[j];
        if (mod === "." || mod === "_" || mod === "-" || mod === "~" || mod === "=") {
            trim =
                mod === "_"
                    ? "ws"
                    : mod === "-"
                      ? "nl"
                      : mod === "~"
                        ? "ws_nl"
                        : mod === "="
                          ? "all"
                          : "none";
            j++;
        }
        const end = this.findDelimiter(j, ce);
        if (end < 0) {
            throw new ParseError("unclosed code block", {
                file: this.file,
                start: i,
                length: cs.length,
                label: "opens here",
            });
        }
        let contentBase = j;
        while (contentBase < end) {
            const c = source.charCodeAt(contentBase);
            if (c === 0x20 || c === 0x09 || c === 0x0a || c === 0x0d) contentBase++;
            else break;
        }
        let contentEnd = end;
        while (contentEnd > contentBase) {
            const c = source.charCodeAt(contentEnd - 1);
            if (c === 0x20 || c === 0x09 || c === 0x0a || c === 0x0d) contentEnd--;
            else break;
        }
        const content = source.slice(contentBase, contentEnd);
        const tokens = lexCode(content, {
            source,
            file: this.file,
            base: contentBase,
        }).filter((t) => t.kind !== "CodeEnd");
        return { tokens, content, contentBase, trim, next: end + ce.length };
    }

    private findDelimiter(from: number, delim: string): number {
        const source = this.source;
        const d0 = delim.charCodeAt(0);
        for (let i = from; i + delim.length <= source.length; i++) {
            if (source.charCodeAt(i) !== d0) continue;
            let ok = true;
            for (let k = 1; k < delim.length; k++) {
                if (source.charCodeAt(i + k) !== delim.charCodeAt(k)) {
                    ok = false;
                    break;
                }
            }
            if (ok) return i;
        }
        return -1;
    }

    private nextMarkup(from: number): number | null {
        const source = this.source;
        const markers = [this.codeStart, this.commentStart];
        let best: number | null = null;
        for (const m of markers) {
            const idx = source.indexOf(m, from);
            if (idx >= 0 && (best === null || idx < best)) best = idx;
        }
        return best;
    }
}

export { parseExpr } from "./expr";
export { ParseError } from "./error";
export { lexCode } from "./lexer";
export type { CodeToken, CodeTokenKind } from "./token";

File: projects/dejavu.py/src/dejavu_language/parser.py
"""Recursive-descent T1 template parser producing Dejavu IR."""

from __future__ import annotations

from typing import Any, Literal

from dejavu_types import DEFAULT_LANGUAGE

from dejavu_language.code_head import (
    CodeHeadExpr,
    CodeHeadIf,
    CodeHeadKind,
    CodeHeadLoop,
    classify_code,
)
from dejavu_language.error import ParseError
from dejavu_language.expr import parse_expr
from dejavu_language.lexer import lex_code

TrimMode = Literal["none", "ws", "nl", "ws_nl", "all"]


def parse_to_ir(source: str, file: str = "template.dejavu") -> dict[str, Any]:
    return {
        "irVersion": "1.0",
        "language": {
            "syntaxMode": DEFAULT_LANGUAGE["syntaxMode"],
            "template": dict(DEFAULT_LANGUAGE["template"]),
        },
        "body": {"type": "Template", "children": TemplateParser(source, file).parse()},
    }


class TemplateParser:
    def __init__(self, source: str, file: str) -> None:
        self.source = source
        self.file = file

    def parse(self) -> list[dict[str, Any]]:
        children, _ = self._parse_body(0, [])
        return children

    def _parse_body(
        self, i: int, stop: list[CodeHeadKind]
    ) -> tuple[list[dict[str, Any]], int]:
        children: list[dict[str, Any]] = []
        source = self.source

        while i < len(source):
            if source.startswith("<%", i) and stop:
                block = self._read_code_block(i)
                head = classify_code(block["content"], block["content_base"], block["tokens"])
                if head.kind in stop:
                    return children, i

            if source.startswith("<#", i):
                end = self._find_delimiter(i + 2, "#>")
                if end < 0:
                    raise ParseError(
                        "unclosed comment",
                        file=self.file,
                        start=i,
                        length=2,
                        label="comment starts here",
                    )
                children.append({"type": "Comment", "value": source[i + 2 : end]})
                i = end + 2
                continue

            if source.startswith("<%!", i):
                children.append({"type": "Text", "value": "<%"})
                i += 3
                continue

            if source.startswith("<%", i):
                open_pos = i
                block = self._read_code_block(i)
                i = block["next"]
                head = classify_code(block["content"], block["content_base"], block["tokens"])

                if stop and head.kind in stop:
                    return children, open_pos

                if head.kind == "if":
                    assert isinstance(head, CodeHeadIf)
                    node, i = self._parse_if(i, head, block["trim"])
                    children.append(node)
                elif head.kind == "loop":
                    assert isinstance(head, CodeHeadLoop)
                    node, i = self._parse_loop(i, head, block["trim"])
                    children.append(node)
                elif head.kind in ("end_if", "end_loop", "else", "else_if"):
                    if not stop:
                        raise ParseError(
                            f"unexpected control `{head.kind}`",
                            file=self.file,
                            start=block["content_base"],
                            length=1,
                        )
                    return children, open_pos
                else:
                    assert isinstance(head, CodeHeadExpr)
                    children.append(
                        {
                            "type": "Interpolation",
                            "expression": parse_expr(
                                head.expr_slice,
                                source=source,
                                file=self.file,
                                base=head.expr_abs,
                            ),
                            "trim": block["trim"],
                        }
                    )
                continue

            nxt = self._next_markup(i)
            if nxt is None:
                children.append({"type": "Text", "value": source[i:]})
                break
            if nxt > i:
                children.append({"type": "Text", "value": source[i:nxt]})
                i = nxt
            else:
                i += 1

        return children, i

    def _parse_if(
        self, i: int, head: CodeHeadIf, trim: TrimMode
    ) -> tuple[dict[str, Any], int]:
        test = parse_expr(
            head.test_slice,
            source=self.source,
            file=self.file,
            base=head.test_abs,
        )
        consequent, i = self._parse_body(i, ["else_if", "else", "end_if"])
        else_ifs: list[dict[str, Any]] = []
        alternate: list[dict[str, Any]] | None = None

        while True:
            block = self._read_code_block(i)
            h = classify_code(block["content"], block["content_base"], block["tokens"])
            if h.kind == "else_if":
                i = block["next"]
                t = parse_expr(
                    h.test_slice,
                    source=self.source,
                    file=self.file,
                    base=h.test_abs,
                )
                body, i = self._parse_body(i, ["else_if", "else", "end_if"])
                else_ifs.append(
                    {
                        "type": "Stmt.ElseIf",
                        "test": t,
                        "consequent": body,
                        "trim": "none",
                    }
                )
            elif h.kind == "else":
                i = block["next"]
                body, i = self._parse_body(i, ["end_if"])
                alternate = body
                end = self._read_code_block(i)
                if (
                    classify_code(end["content"], end["content_base"], end["tokens"]).kind
                    != "end_if"
                ):
                    raise ParseError(
                        "expected `end if`",
                        file=self.file,
                        start=end["content_base"],
                        length=1,
                    )
                i = end["next"]
                break
            elif h.kind == "end_if":
                i = block["next"]
                break
            else:
                raise ParseError(
                    f"expected if closer, got `{h.kind}`",
                    file=self.file,
                    start=block["content_base"],
                    length=1,
                )

        node: dict[str, Any] = {
            "type": "Stmt.If",
            "test": test,
            "consequent": consequent,
            "elseIfs": else_ifs,
            "trim": trim,
        }
        if alternate is not None:
            node["alternate"] = alternate
        return node, i

    def _parse_loop(
        self, i: int, head: CodeHeadLoop, trim: TrimMode
    ) -> tuple[dict[str, Any], int]:
        iterable = parse_expr(
            head.iter_slice,
            source=self.source,
            file=self.file,
            base=head.iter_abs,
        )
        body, i = self._parse_body(i, ["end_loop"])
        end = self._read_code_block(i)
        if (
            classify_code(end["content"], end["content_base"], end["tokens"]).kind
            != "end_loop"
        ):
            raise ParseError(
                "expected `end loop`",
                file=self.file,
                start=end["content_base"],
                length=1,
            )
        return (
            {
                "type": "Stmt.For",
                "item": head.item,
                "iterable": iterable,
                "body": body,
                "trim": trim,
            },
            end["next"],
        )

    def _read_code_block(self, i: int) -> dict[str, Any]:
        source = self.source
        if not source.startswith("<%", i):
            raise ParseError(
                "expected code open `<%`",
                file=self.file,
                start=i,
                length=1,
            )
        j = i + 2
        trim: TrimMode = "none"
        if j < len(source) and source[j] in "._-~=":
            mod = source[j]
            trim = {
                ".": "none",
                "_": "ws",
                "-": "nl",
                "~": "ws_nl",
                "=": "all",
            }[mod]
            j += 1
        end = self._find_delimiter(j, "%>")
        if end < 0:
            raise ParseError(
                "unclosed code block",
                file=self.file,
                start=i,
                length=2,
                label="opens here",
            )
        content_base = j
        while content_base < end:
            c = ord(source[content_base])
            if c in (0x20, 0x09, 0x0A, 0x0D):
                content_base += 1
            else:
                break
        content_end = end
        while content_end > content_base:
            c = ord(source[content_end - 1])
            if c in (0x20, 0x09, 0x0A, 0x0D):
                content_end -= 1
            else:
                break
        content = source[content_base:content_end]
        tokens = [
            t
            for t in lex_code(content, source=source, file=self.file, base=content_base)
            if t.kind != "CodeEnd"
        ]
        return {
            "tokens": tokens,
            "content": content,
            "content_base": content_base,
            "trim": trim,
            "next": end + 2,
        }

    def _find_delimiter(self, from_pos: int, delim: str) -> int:
        source = self.source
        d0 = ord(delim[0])
        dlen = len(delim)
        i = from_pos
        limit = len(source) - dlen + 1
        while i < limit:
            if ord(source[i]) != d0:
                i += 1
                continue
            ok = True
            for k in range(1, dlen):
                if ord(source[i + k]) != ord(delim[k]):
                    ok = False
                    break
            if ok:
                return i
            i += 1
        return -1

    def _next_markup(self, from_pos: int) -> int | None:
        source = self.source
        i = from_pos
        n = len(source)
        while i + 1 < n:
            if ord(source[i]) != 0x3C:  # <
                i += 1
                continue
            nxt = ord(source[i + 1])
            if nxt in (0x25, 0x23):  # % or #
                return i
            i += 1
        return None
